src/phoenix/experimental/evals/models/rate_limiters.py

```diff
@@ -0,0 +1,193 @@
+            raise UnavailableTokensError
+        self.tokens -= 1
+
+    def wait_until_ready(
+        self,
+        max_wait_time: float = 300,
+    ) -> None:
+        start = time.time()
+        while (time.time() - start) < max_wait_time:
+            try:
+                self.increase_rate()
+                self.make_request_if_ready()
+                break
+            except UnavailableTokensError:
+                time.sleep(0.1 / self.rate)
+                continue
+
+    async def async_wait_until_ready(
+        self,
+        max_wait_time: float = 300,
+    ) -> None:
+        start = time.time()
+        while (time.time() - start) < max_wait_time:
+            try:
+                self.increase_rate()
+                self.make_request_if_ready()
+                break
+            except UnavailableTokensError:
+                await asyncio.sleep(0.1 / self.rate)
+                continue
+
+
+class RateLimiter:
+    def __init__(
+        self,
+        rate_limit_error: Type[BaseException],
+        max_rate_limit_retries: int = 10,
+        initial_per_second_request_rate: float = 1,
+        maximum_per_second_request_rate: float = 50,
+        enforcement_window_minutes: float = 1,
+        rate_reduction_factor: float = 0.5,
+        rate_increase_factor: float = 0.01,
+        cooldown_seconds: float = 5,
+    ) -> None:
+        self._rate_limit_error = rate_limit_error
+        self._max_rate_limit_retries = max_rate_limit_retries
+        self._throttler = AdaptiveTokenBucket(
+            initial_per_second_request_rate=initial_per_second_request_rate,
+            maximum_per_second_request_rate=maximum_per_second_request_rate,
+            enforcement_window_minutes=enforcement_window_minutes,
+            rate_reduction_factor=rate_reduction_factor,
+            rate_increase_factor=rate_increase_factor,
+            cooldown_seconds=cooldown_seconds,
+        )
+
+    def limit(
+        self, fn: Callable[ParameterSpec, GenericType]
+    ) -> Callable[ParameterSpec, GenericType]:
+        @wraps(fn)
+        def wrapper(*args: Any, **kwargs: Any) -> GenericType:
+            try:
+                self._throttler.wait_until_ready()
+                request_start_time = time.time()
+                return cast(GenericType, fn(*args, **kwargs))  # type: ignore
+            except self._rate_limit_error:
+                self._throttler.on_rate_limit_error(request_start_time)
+                for _attempt in range(self._max_rate_limit_retries):
+                    try:
+                        request_start_time = time.time()
+                        self._throttler.wait_until_ready()
+                        return cast(GenericType, fn(*args, **kwargs))  # type: ignore
+                    except self._rate_limit_error:
+                        self._throttler.on_rate_limit_error(request_start_time)
+                        continue
+            raise self._rate_limit_error("Exceeded max retries")
```

Comment:
I generally agree that I prefer to avoid nesting, but here I want to keep the rate-limit handling logic separate from the initial attempt, despite a lot of the code being repeated.